server/paymentProviders/stripe/virtual-cards.js

```diff
@@ -51,3 +58,173 @@ export const assignCardToCollective = async (cardNumber, expireDate, cvv, collec
 
   return await models.VirtualCard.create(cardData);
 };
+
+export const processTransaction = async (stripeEvent, stripeSignature) => {
+  const stripeTransaction = stripeEvent.data.object;
+
+  const virtualCard = await models.VirtualCard.findOne({
+    where: {
+      id: stripeTransaction.card,
+    },
+    include: [
+      { association: 'collective', required: true },
+      { association: 'host', required: true },
+      { association: 'user' },
+    ],
+  });
+
+  if (!virtualCard) {
+    logger.error('stripe/webhook: could not find VirtualCard', { body: stripeTransaction });
+    return;
+  }
+
+  const host = virtualCard.host;
+  const collective = virtualCard.collective;
+  const [connectedAccount] = await host.getConnectedAccounts({ where: { service: 'stripe' } });
+
+  if (!connectedAccount) {
+    logger.error('stripe/webhook: host is not connected to Stripe', { body: stripeTransaction });
+    return;
+  }
+
+  const stripe = Stripe(connectedAccount.token);
+
+  try {
+    stripe.webhooks.constructEvent(stripeEvent, stripeSignature, connectedAccount.data.stripeEndpointSecret);
+  } catch {
+    logger.error('stripe/webhook: source of event not recognized', { body: stripeTransaction });
+    return;
+  }
+
+  const amount = stripeTransaction.amount;
+
+  if (amount === 0) {
+    return;
+  }
+
+  const existingExpense = await models.Expense.findOne({
+    where: {
+      VirtualCardId: virtualCard.id,
+      data: { token: stripeTransaction.id },
+    },
+  });
+  if (existingExpense) {
+    logger.warn(`Virtual Card charge already reconciled, ignoring it: ${stripeTransaction.id}`);
+    return;
+  }
+
+  const UserId = virtualCard.UserId;
+  const isRefund = amount < 0;
+
+  // If it is refund, we'll check if the transaction was already created because there are no expenses created for refunds.
+  if (isRefund) {
+    const existingTransaction = await models.Transaction.findOne({
+      where: {
+        CollectiveId: collective.id,
+        data: { token: stripeTransaction.id },
+      },
+    });
+    if (existingTransaction) {
+      logger.warn(`Virtual Card refund already reconciled, ignoring it: ${stripeTransaction.id}`);
+      return;
+    }
+  }
+
+  let expense;
+  try {
+    const slug = toString(stripeTransaction['merchant_data']['network_id']).toLowerCase();
+    const [vendor] = await models.Collective.findOrCreate({
+      where: { slug },
+      defaults: { name: stripeTransaction['merchant_data']['name'], type: CollectiveTypes.VENDOR },
+    });
+
+    const hostCurrencyFxRate = await getFxRate('USD', host.currency);
```

Comment:
Do we assume that all incoming transactions will be in dollars? Should we already prepare for the case where it's not?

Reply:
Transaction are in the currency of the card, right now we can only have USD cards. I think we should extend this behavior for other currency but in other PR.

Reply:
Are you confident that transaction made in EUR for those cards will come back to us in USD if the card is USD?

Anyway, no problem, this can be handled later

Reply:
Yes I am confident.